pkg/k8s/resource/resource.go

```diff
@@ -522,100 +447,266 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
-			kq.ShutDown()
+			s.wq.ShutDown()
 		case ErrorActionIgnore:
-			kq.Forget(entry)
+			s.wq.Forget(entry)
 		default:
-			panic(fmt.Sprintf("keyQueue: unknown action %q from error handler %v", action, kq.errorHandler))
+			panic(fmt.Sprintf("keyQueue: unknown action %q from error handler %v", action, s.options.errorHandler))
 		}
 	} else {
 		// As the object was processed successfully we can "forget" it.
 		// This clears any rate limiter state associated with this object, so
 		// it won't be throttled based on previous failure history.
-		kq.Forget(entry)
+		s.wq.Forget(entry)
 	}
 }
 
-// queueEntry restricts the set of types we use when type-switching over the
+// workItem restricts the set of types we use when type-switching over the
 // queue entries, so that we'll get a compiler error on impossible types.
 //
 // The queue entries must be kept comparable and not be pointers as we want
-// to be able to coalesce multiple upsertEntry's into a single element in the
+// to be able to coalesce multiple keyEntry's into a single element in the
 // queue.
-type queueEntry interface {
-	isQueueEntry()
+type workItem interface {
+	isWorkItem()
 }
 
-type syncEntry struct{}
+// syncWorkItem marks the store as synchronized and thus a 'Sync' event can be
+// emitted to the subscriber.
+type syncWorkItem struct{}
 
-func (syncEntry) isQueueEntry() {}
+func (syncWorkItem) isWorkItem() {}
 
-type upsertEntry struct {
+// keyWorkItem marks work for a specific key. Whether this is an upsert or delete
+// depends on the state of the store at the time this work item is processed.
+type keyWorkItem struct {
 	key Key
 }
 
-func (upsertEntry) isQueueEntry() {}
+func (keyWorkItem) isWorkItem() {}
+
+func (r *resource[T]) newInformer() (cache.Indexer, cache.Controller) {
+	clientState := cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc, r.opts.indexers)
+	opts := cache.DeltaFIFOOptions{KeyFunction: cache.MetaNamespaceKeyFunc, KnownObjects: clientState}
+	fifo := cache.NewDeltaFIFOWithOptions(opts)
+	cacheMutationDetector := cache.NewCacheMutationDetector(fmt.Sprintf("%T", r))
+	transformer := r.opts.transform
+	cfg := &cache.Config{
+		Queue:            fifo,
+		ListerWatcher:    r.lw,
+		ObjectType:       r.opts.sourceObj(),
+		FullResyncPeriod: 0,
+		RetryOnError:     false,
+		Process: func(obj interface{}, isInInitialList bool) error {
+			// Processing of the deltas is done under the resource mutex. This
+			// avoids emitting double events for new subscribers that list the
+			// keys in the store.
+			r.mu.RLock()
+			defer r.mu.RUnlock()
+
+			for _, d := range obj.(cache.Deltas) {
+				var obj interface{}
+				if transformer != nil {
+					var err error
+					if obj, err = transformer(d.Object); err != nil {
+						return err
+					}
+				} else {
+					obj = d.Object
+				}
 
-type deleteEntry struct {
-	key Key
-	obj any
+				// In CI we detect if the objects were modified and panic
+				// this is a no-op in production environments.
```

Comment:
Is this actually still true, or just copy pasta?

Reply:
```
test/helpers/kubectl.go:                // "extraEnv[0].name":              "KUBE_CACHE_MUTATION_DETECTOR",
```

https://github.com/kubernetes/client-go/blob/v0.28.1/tools/cache/mutation_detector.go#L36

Heh, looks like we no longer set it anywhere, so no we don't check in CI.  And having it enabled by default is a fairly big hammer. I'll drop this from this fix as thinking about where and how to enable this is not trivial.